Add rlwe plain ops

johnbntn opened the pull request:
Added lwe ciphertext-plaintext ops and their lowerings for bgv/ckks  (#1603)  

Also I can't fully test the `lwe-to-polynomial` lowering until #1199 but I did my best to add the functionality and tests.

j2kun commented on lib/Dialect/LWE/Conversions/LWEToOpenfhe/LWEToOpenfhe.cpp:
```diff
@@ -400,16 +400,13 @@ struct LWEToOpenfhe : public impl::LWEToOpenfheBase<LWEToOpenfhe> {
         // SubPlain
-        ConvertCiphertextPlaintextOp<bgv::SubPlainOp, openfhe::SubPlainOp>,
-        ConvertCiphertextPlaintextOp<ckks::SubPlainOp, openfhe::SubPlainOp>,
+        ConvertCiphertextPlaintextOp<lwe::RSubOp_Plain, openfhe::SubPlainOp>,
```
You have defined sub_plain in LWE to allow the plaintext in either operand, but IIRC OpenFHE requires it is only in one operand. Could you double-check this, and then maybe this needs a custom rule or canonicalization pattern to ensure the plaintext operand is in the right place?

ZenithalHourlyRate replied:
BGV_SubPlainOp also allows pt in either operand, openfhe has two overloads for `cc->EvalSub` that accepts both of them. It is Lattigo that only accepts `evaluator.Sub(ct, pt)`, but that reordering is handled in `LWEToLattigo` by `Add(Negate(ct), pt)`, where Negate is only present in `lattigo` dialect, so it should not be a concern for `lwe` dialect.

j2kun commented on lib/Dialect/LWE/Conversions/LWEToPolynomial/LWEToPolynomial.cpp:
```diff
@@ -358,6 +358,21 @@ struct ConvertRAdd : public OpConversionPattern<RAddOp> {
+      RAddOp_Plain op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    rewriter.replaceOpWithNewOp<::mlir::heir::polynomial::AddOp>(
+        op, adaptor.getOperands()[0], adaptor.getOperands()[1]);
```
not sure why the other patterns use `getOperands()[0]` in this file, but the `OpAdaptor` should have named methods corresponding to the tablegen names (`getLhs()`, `getRhs()`) that you can use here.

j2kun replied:
in an OpConversionPattern `OpAdaptor` is defined as `SourceOp::Adaptor` where `SourceOp` is the template argument (`RAddOp_Plain` here). Cf. https://github.com/llvm/llvm-project/blob/5dc3cd0ee40c00d9fb542488fa5a54ff70273112/mlir/include/mlir/Transforms/DialectConversion.h#L586

AlexanderViand replied:
Oh, I see - I guess I was thinking of the (now removed) OneToNConversionPattern case, where it's really just a GenericAdaptor... that *does* raise the question (unrelated to this PR) of why all the patterns here use the same style 🙈

j2kun commented on lib/Dialect/LWE/Conversions/LWEToPolynomial/LWEToPolynomial.cpp:
```diff
@@ -467,6 +497,52 @@ struct ConvertRMul : public OpConversionPattern<RMulOp> {
+    auto y = adaptor.getRhs();
+    auto yT = cast<RankedTensorType>(y.getType());
+
+    if (xT.getNumElements() != 2 || yT.getNumElements() != 1) {
```
If the plaintext op can be in either operand, you may want to add a step before this that checks the types of `op.getLhs()` and `op.getRhs()` and then swaps `x` and `y` appropriately.

Or we could:

- make the tablegen force the LHS to be ciphertext and RHS to be plaintext
- add a general canonicalization rule moves the plaintext to the RHS. 

The lowerings to LWE may already ensure this, but new lowerings to LWE may not be aware of this requirement.

j2kun commented on lib/Dialect/LWE/Conversions/LWEToPolynomial/LWEToPolynomial.cpp:
```diff
@@ -467,6 +497,52 @@ struct ConvertRMul : public OpConversionPattern<RMulOp> {
+    ImplicitLocOpBuilder b(op->getLoc(), rewriter);
+    // z = mul([x0, x1], [y0]) := [x0y0, x1y0] (Multiply ciphertext [2dim] with
+    // plaintext [1dim]) bgv and ckks canonicalize with ciphertext first
+    auto i0 = b.create<arith::ConstantIndexOp>(0);
+    auto i1 = b.create<arith::ConstantIndexOp>(1);
+
+    auto x0 =
+        b.create<tensor::ExtractOp>(xT.getElementType(), x, ValueRange{i0});
+    auto x1 =
+        b.create<tensor::ExtractOp>(xT.getElementType(), x, ValueRange{i1});
+
+    auto y0 =
+        b.create<tensor::ExtractOp>(yT.getElementType(), y, ValueRange{i0});
+
+    auto z0 = b.create<::mlir::heir::polynomial::MulOp>(x0, y0);
+    auto z1 = b.create<::mlir::heir::polynomial::MulOp>(x1, y0);
+
+    auto z = b.create<tensor::FromElementsOp>(ArrayRef<Value>({z0, z1}));
```
`polynomial::MulOp` is supposed to also allow tensor-typed inputs and apply mul elementwise. If you weren't aware of that, you could try simplifying this pattern by doing

```
auto repeated = b.create<tensor::FromElementsOp>({y, y});
auto prod  = b.create<polynomial::MulOp>(x, repeated);
```

j2kun reviewed (commented):
Generally looks good! I have some suggested simplifications and then I would leave time for @ZenithalHourlyRate or @AlexanderViand to comment before final approval.

AlexanderViand reviewed (changes requested):
As Jeremy already pointed out, some backends restrict which side of the op the plaintext can appear on. This is most annoying for `sub_plain`. I think @ZenithalHourlyRate knows more on what, if any, restrictions there really are, though.

Even if it turns out not to be an issue, I think all the tests, including the various `<scheme>_to_<backend>.mlir` deserve to have a pt-lhs and a pt-rhs version of each op, just to make sure this is explicitly tested for.

j2kun commented on MODULE.bazel.lock:
```diff
@@ -1610,7 +1610,7 @@
     "@@rules_python+//python/extensions:pip.bzl%pip": {
       "general": {
-        "bzlTransitiveDigest": "I2tEsg/D0/3B4CIPYr4NzeUfrj2ODZtxU4zM8Z7TzJU=",
+        "bzlTransitiveDigest": "vKDNEc7MJLQpnsEBRqgLsrNo+zskN5XVJzAg1j5LtPs=",
```
I think if you rebase this will go away. It is fine to commit changes to the lockfile though. We will eventually set up something that does this periodically so it doesn't show up in random PRs.

ZenithalHourlyRate commented on lib/Dialect/LWE/IR/LWEOps.td:
```diff
@@ -159,13 +165,21 @@ def LWE_RAddOp : LWE_BinOp<"radd", [SameOperandsAndResultRings, SameOperandsAndR
   }];
 }
 
+def LWE_RAddOp_Plain : LWE_CiphertextPlaintextOp<"radd_plain", [Commutative, SameOperandsAndResultPlaintextTypes, InferTypeOpAdaptor]> {
```
```suggestion
def LWE_RAddPlainOp : LWE_CiphertextPlaintextOp<"radd_plain", [Commutative, SameOperandsAndResultPlaintextTypes, InferTypeOpAdaptor]> {
```

In HEIR we prefer a name like `LWE_RAddPlainOp` so the cpp op name will become `lwe::RAddPlainOp` instead of `lwe::RAddOp_Plain`. I do not think we have op name that has underscore in it. A mnemonic with underscore is accepted.

ZenithalHourlyRate replied:
Also, please add canonicalizer on ct-pt ordering to LWE_RAddPlainOp and LWE_RMulPlainOp, it is just copy-pasting what 948dd4f66 has done in BGVOps.td and BGVOps.cpp.

ZenithalHourlyRate commented on lib/Dialect/LWE/Conversions/LWEToLattigo/LWEToLattigo.cpp:
ConvertRlweCommutativePlainOp and ConvertRlweSubPlainOp have already taken care of it so it should not be a concern for at `lwe` dialect level.

ZenithalHourlyRate reviewed (changes requested):
Generally LGTM! Just some op naming convention suggestion.

About the ct-pt ordering problem, openfhe/lattigo lowerings for lwe already handle them so it should not be an issue for lwe dialect itself. I think the polynomial lowering should be aware of it instead of enforcing lwe dialect to have some constraint.

ZenithalHourlyRate commented:
> For some reason when I was adding bgv/ckks-to-openfhe tests I continued to fail them even though my test log showed the expected output. When I added -dump-input=help as a FileCheck flag they passed. Has anyone else ever seen this behavior?

No, it should mean the CHECK directive you wrote does not match the input.

Also, according to https://llvm.org/docs/CommandGuide/FileCheck.html#cmdoption-FileCheck-dump-input, `-dump-input=help` means `Explain input dump and quit`, so it should not fail. And bazel test asserts PASSED or not by checking the return value of executable, so you will always get a PASSED with such flag.

johnbntn commented:
I think I added all of the requested changes:
- Removed underscore in op names
- Added canonicalizers for add/mul plain
- Simplified the poly lowering
- Fixed the openfhe lowering tests by pattern matching just the plaintext instead of the plaintext and return types
- Added some more tests for orderings and canonicalizers

Please let me know if I missed anything or if you have any more requested changes

ZenithalHourlyRate reviewed (approved):
LGTM!

There are still some comments not addressed

https://github.com/google/heir/pull/1776#discussion_r2069310687

on usage of adaptor

https://github.com/google/heir/pull/1776#discussion_r2069593135

on `rlwe_ops.mlir` containing IR syntax test for pt-ct. I know there are some tests on them on the lowering part (e.g. bgv-to-openfhe), but the IR syntax test itself should also reflect that.

ZenithalHourlyRate commented:
Also remember to squash the commits so that the PR merge tool can pick it up.

j2kun commented:
> I believe I finished making all of the requested changes but please let me know if I missed anything.

Due to our strange project setup in re Google, you also need to squash your commits before we can merge it.
